---
title: "2023_06_01 DAA"
author: "kim soyeon"
date: "2023-06-01"
output: html_document
---

```{r}

library(phyloseq)
library(tidyverse)
library(data.table)
library(dplyr)
library(reshape2)
library(ggplot2)
library(stringr)



ps <- readRDS("./ps.rds") ###################################################
ps.gt <-  subset_samples(ps, body.site %in% c("gut", "tongue"))
ps.gt.g <- tax_glom(ps.gt, taxrank = "Genus")


```



# 2. DA test_microbiomeMarker


## lefse 
```{r}
library(microbiomeMarker)
```




```{r}

ps.gt.rel <- transform_sample_counts(ps.gt.g, function(x) x/sum(x) )

lefse <- run_lefse(
  ps.gt.rel,
  wilcoxon_cutoff = 0.05,
  group = "body.site",
  kw_cutoff = 0.05,
  taxa_rank = "Genus",
  multigrp_strat = TRUE,
  lda_cutoff = 2,
  bootstrap_n = 999
)

marker_table(lefse)
microbiomeMarker::plot_ef_bar(lefse)
  ggsave("./lefse_result.png", device = "png",dpi =  300)
lefse_result <- marker_table(lefse)$feature
lefse_result

```



## 2) edge_R # OK
```{r}


edgeR <- run_edger(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    p_adjust = "fdr",
    taxa_rank = "Genus",
    norm = "TMM"
)
marker_table(edgeR)
microbiomeMarker::plot_ef_bar(edgeR)
  ggsave("./edgeR_result.png", device = "png",dpi =  300)


```

## 3) DESeq2 = Script VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV
```{r}

DESeq2 <- run_deseq2(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    p_adjust = "fdr",
    taxa_rank = "Genus",
    norm = "RLE"
)
marker_table(DESeq2)
plot_ef_bar(DESeq2)


```
## 4) limma_voom = Script VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV
```{r}
Limma_voom <- run_limma_voom(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    p_adjust = "fdr",
    taxa_rank = "Genus",
    norm = "TMM"
)
marker_table(Limma_voom)
plot_ef_bar(Limma_voom)

```
## 5) metagenomeSeq # OK
```{r}
metagenomeseq <- run_metagenomeseq(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    p_adjust = "fdr",
    taxa_rank = "Genus",
    norm = "CSS",
    transform  = "log10"
)
marker_table(metagenomeseq)
plot_ef_bar(metagenomeseq)



```


## 6) ALDEx2  ####################### Error
```{r}

ps.gt.rel <- transform_sample_counts(ps.gt, function(x) x/sum(x))


aldex2 <- run_aldex(
    ps.gt.rel,
    group = "body.site",
    # pvalue_cutoff = 0.05,
    # method ="wilcox.test",
    # p_adjust = "fdr",
    # taxa_rank = "Genus",
    # norm = "CLR"
)
marker_table(aldex2)
plot_ef_bar(aldex2)

 



```


## 7) ANCOM-2 ####################### Error?
```{r}


ancom <- run_ancom(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    p_adjust = "fdr",
    taxa_rank = "Genus",
    transform = "identity",
    norm = "TSS"
)
marker_table(ancom)
plot_ef_bar(ancom)




```

## 8) ancombc # No marker 
```{r}



ancombc <- run_ancombc(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    p_adjust = "fdr",
    taxa_rank = "Genus",
)
marker_table(ancombc)
plot_ef_bar(ancombc)



```

## 9) supervised learning (SL)
```{r}
?run_sl

LR <- run_sl(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    method = "LR",
    p_adjust = "fdr",
    taxa_rank = "Genus",
)
marker_table(ancombc)
plot_sl_roc(LR, group = "Group")



RF <- run_sl(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    method = "RF",
    p_adjust = "fdr",
    taxa_rank = "Genus",
) #  No importance values available

SVM <- run_sl(
    ps.gt,
    group = "body.site",
    pvalue_cutoff = 0.05,
    method = "SVM",
    p_adjust = "fdr",
    taxa_rank = "Genus"
)  # constant. Cannot scale data

```


# 4. DA test_Nicholas Ollberding
- 출처 : https://www.nicholas-ollberding.com/post/identifying-differentially-abundant-features-in-microbiome-data/
```{r}
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# 
# BiocManager::install("curatedMetagenomicData")

library(curatedMetagenomicData); packageVersion("curatedMetagenomicData")   
## [1] '1.18.2'
library(tidyverse); packageVersion("tidyverse")                             
## [1] '1.3.0'
library(phyloseq); packageVersion("phyloseq")                               
## [1] '1.32.0'
library(edgeR); packageVersion("edgeR")                                     
## [1] '3.30.3'
library(limma); packageVersion("limma")                                    
## [1] '3.44.3'
# BiocManager::install("DEFormats")
library(DEFormats); packageVersion("DEFormats")                             
## [1] '1.18.0'
library(DESeq2); packageVersion("DESeq2")                                   
## [1] '1.28.1'
library(apeglm); packageVersion("apeglm")                                   
## [1] '1.10.0'
library(corncob); packageVersion("corncob")                                 
## [1] '0.1.0'
library(ANCOMBC); packageVersion("ANCOMBC")                                 
## [1] '1.0.3'
library(Maaslin2); packageVersion("Maaslin2")                               
## [1] '1.2.0'
library(VennDiagram); packageVersion("VennDiagram")                         
## [1] '1.6.20'

# library(curatedMetagenomicData); packageVersion("curatedMetagenomicData")   
# dhakan_eset <- DhakanDB_2019.metaphlan_bugs_list.stool()
# a <- curatedMetagenomicData::curatedMetagenomicData()

```



## 1) Limma-Voom with TMM normalization = microbiomeMarker VVVVVVVVVVVVVVVVVVVVV
```{r}
Oral_deseq <- phyloseq_to_deseq2(oral.sp2, ~ Group)      #convert to DESeq2 and DGEList objects
Oral_deseq.list <- as.DGEList(Oral_deseq)
Oral_deseq.TMM <- calcNormFactors(Oral_deseq.list, method = "TMM")
  #store TMM norm factor
?calcNormFactors
head(Oral_deseq.TMM$samples$norm.factors)


mm <- model.matrix(~ group, Oral_deseq.TMM$samples)              #construct model matrix          
head(mm)

table(mm[, 2])

y <- voom(Oral_deseq.TMM, mm, plot = T)                          #obtain Voom weights


fit <- lmFit(y, mm)                                   #fit lm with limma
fit <- eBayes(fit)
head(coef(fit))

limma_res_df <- data.frame(topTable(fit, coef = "groupHV", number = Inf))    #extract results
limma_res_df["Porphyromonas_gingivalis(AB035459)", ]
# P.Value     0.004472435
# adj.P.Val   0.2269345	

fdr_limma <- limma_res_df %>%
    dplyr::filter(adj.P.Val < 0.05) %>%
    rownames_to_column(var = "Species")
    # Johnsonella_ignava(X87152)

dim(fdr_limma)

fdr_limma2 <- fdr_limma 


limma_result_Xfdr <- limma_res_df %>%
  dplyr::filter(P.Value  <= 0.05) %>%
  rownames_to_column(var = "Species")  %>%  .$Species
 

```

```{r}

Oral.kg_deseq <- phyloseq_to_deseq2(oral.sp2.kg, ~ Group)      #convert to DESeq2 and DGEList 
Oral.kg_deseq.list <- as.DGEList(Oral.kg_deseq)
Oral.kg_deseq.TMM <- calcNormFactors(Oral.kg_deseq.list, method = "TMM") # store TMM norm factor
mm.kg <- model.matrix(~ group, Oral.kg_deseq.TMM$samples) # construct model matrix   
y.kg <- voom(Oral.kg_deseq.TMM, mm.kg, plot = T) # obtain Voom weights


?lmFit
fit.kg <- lmFit(y.kg, mm.kg) # fit lm with limma
fit.kg <- eBayes(fit.kg)
head(coef(fit.kg))

limma_res_df.kg <- data.frame(topTable(fit.kg, coef = "groupHV", number = Inf)) 
limma_res_df.kg["Porphyromonas_gingivalis(AB035459)", ]
# P.Value     0.03203517	
# adj.P.Val   0.9373918	

fdr_limma.kg <- limma_res_df.kg %>%
    dplyr::filter(adj.P.Val < 0.05) %>%
    rownames_to_column(var = "Species")
    # Johnsonella_ignava(X87152)
fdr_limma.kg # XXXXXXXXX

```

```{r}
Oral.bu_deseq <- phyloseq_to_deseq2(oral.sp2.bu, ~ Group)      #convert to DESeq2 and DGEList 
Oral.bu_deseq.list <- as.DGEList(Oral.bu_deseq)
Oral.bu_deseq.TMM <- calcNormFactors(Oral.bu_deseq.list, method = "TMM") # store TMM norm factor
mm.bu <- model.matrix(~ group, Oral.bu_deseq.TMM$samples) # construct model matrix   
y.bu <- voom(Oral.bu_deseq.TMM, mm.bu, plot = T) # obtain Voom weights

fit.bu <- lmFit(y.bu, mm.bu) # fit lm with limma
fit.bu <- eBayes(fit.bu)
head(coef(fit.bu))

limma_res_df.bu <- data.frame(topTable(fit.bu, coef = "groupHV", number = Inf)) 
limma_res_df.bu["Porphyromonas_gingivalis(AB035459)", ]
# P.Value     0.02052124	
# adj.P.Val   0.3775914


fdr_limma.bu <- limma_res_df.bu %>%
    dplyr::filter(adj.P.Val < 0.05) %>%
    rownames_to_column(var = "Species")
    # Johnsonella_ignava(X87152)
fdr_limma.bu # XXXXXXXXX

```



## 2) DESeq2 with apeglm  = microbiomeMarker VVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVVV


```{r}
oral.sp2
Oral_deseq <- phyloseq_to_deseq2(oral.sp2, ~ Group)      #convert to DESeq2 and DGEList 
Oral_deseq2 <- DESeq(Oral_deseq, test = "Wald", 
                     fitType = "local", sfType = "poscounts")
plotDispEsts(Oral_deseq2)

res <- lfcShrink(Oral_deseq2, coef=2, type="apeglm")

plotMA(Oral_deseq2)

deseq_res_df <- data.frame(res) %>%
  rownames_to_column(var = "Species") %>%
  dplyr::arrange(padj)                                 

fdr_deseq <- deseq_res_df %>%
    dplyr::filter(padj < 0.05) 
dim(fdr_deseq) # 49 6
head(fdr_deseq)
fdr_deseq[fdr_deseq$Species == "Porphyromonas_gingivalis(AB035459)", ]


deseq_res_df %>%
    dplyr::filter(pvalue <= 0.05) %>% 
  arrange(Species)

# ggplot(fdr_deseq, aes(x = Species, y = log2FoldChange, color = Species)) +
#     geom_point(size = 4) +
#     labs(y = "\nLog2 Fold-Change for AA vs. HV", x = "") +
#     theme(axis.text.x = element_text(color = "black", size = 12),
#           axis.text.y = element_text(color = "black", size = 12),
#           axis.title.y = element_text(size = 14),
#           axis.title.x = element_text(size = 14),
#           legend.text = element_text(size = 12),
#           legend.title = element_text(size = 12),
#           legend.position = "none") +
#     coord_flip() +
#     geom_hline(yintercept = 0, linetype="dotted")

```

```{r}

oral.kg_deseq <- phyloseq_to_deseq2(oral.sp2.kg, ~ Group)      #convert to DESeq2 and DGEList 
Oral.kg_deseq2 <- DESeq(oral.kg_deseq, test = "Wald", 
                     fitType = "local", sfType = "poscounts")
plotDispEsts(Oral.kg_deseq2)
res.kg <- lfcShrink(Oral.kg_deseq2, coef=2, type="apeglm")
plotMA(Oral.kg_deseq2)

deseq_res_df.kg <- data.frame(res.kg) %>%
  rownames_to_column(var = "Species") %>%
  dplyr::arrange(padj)                                 

fdr_deseq.kg <- deseq_res_df.kg %>%
    dplyr::filter(padj < 0.05) 
dim(fdr_deseq.kg) #  4 6
head(fdr_deseq.kg)


```

```{r}

oral.bu_deseq <- phyloseq_to_deseq2(oral.sp2.bu, ~ Group)      #convert to DESeq2 and DGEList 
Oral.bu_deseq2 <- DESeq(oral.bu_deseq, test = "Wald", 
                     fitType = "local", sfType = "poscounts")
plotDispEsts(Oral.bu_deseq2)
res.bu <- lfcShrink(Oral.bu_deseq2, coef=2, type="apeglm")
plotMA(Oral.bu_deseq2)

deseq_res_df.bu <- data.frame(res.bu) %>%
  rownames_to_column(var = "Species") %>%
  dplyr::arrange(padj)                                 

fdr_deseq.bu <- deseq_res_df.bu %>%
    dplyr::filter(padj < 0.05) 
dim(fdr_deseq.bu) #  19 6
head(fdr_deseq.bu)


```




## 3) Corncob : pg OK
```{r}
corn_da <- differentialTest(formula = ~ Group,
                            phi.formula = ~ 1,
                            formula_null = ~ 1,
                            phi.formula_null = ~ 1,
                            data = oral.sp2,
                            test = "Wald", boot = FALSE,
                            fdr_cutoff = 0.05)
corn_da_p <-corn_da$p 
Corncob_result_Xfdr <- corn_da_p[corn_da_p<=0.05] %>% names %>% sort

corn_da_q <-corn_da$p_fdr
corn_da_q[corn_da_q<=0.05] %>% names %>% sort


fdr_corncob <- corn_da$significant_taxa

fdr_corncob <- fdr_corncob %>%
  dplyr::filter(q_val < 0.05)


Corncob_result_Xfdr <- ancom_res_df %>%
  dplyr::filter(p_val < 0.05) %>% 
  arrange(Species) %>% .%Species


dim(data.frame(fdr_corncob))

result <- sort(corn_da$p_fdr) %>% as.data.frame()#  737
corn_da$p_fdr < 0.05 %>%  table




```

```{r}
corn_da.kg <- differentialTest(formula = ~ Group,
                            phi.formula = ~ 1,
                            formula_null = ~ 1,
                            phi.formula_null = ~ 1,
                            data = oral.sp2.kg,
                            test = "Wald", boot = FALSE,
                            fdr_cutoff = 0.05)

fdr_corncob.kg <- corn_da.kg$significant_taxa

fdr_ancom.kg <- fdr_corncob.kg %>%
  dplyr::filter(q_val < 0.05)

```
```{r}
corn_da.bu <- differentialTest(formula = ~ Group,
                            phi.formula = ~ 1,
                            formula_null = ~ 1,
                            phi.formula_null = ~ 1,
                            data = oral.sp2.bu,
                            test = "Wald", boot = FALSE,
                            fdr_cutoff = 0.05)
corn_da.bu$p_fdr %>%   sort() 
fdr_corncob.bu <- corn_da.bu$significant_taxa

```


## 4) MaAsLin2 : pg OK but 보정 X
```{r}

mas_1 <- Maaslin2(
  input_data = data.frame(otu_table(oral.sp2)),
  input_metadata = data.frame(sample_data(oral.sp2)),
  output = "./output/5.DA_decontam/etc/Maaslin2_1/",
  min_abundance = 0.0,
  min_prevalence = 0.0,
  normalization = "TSS",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "Group",
  correction = "BH",
  standardize = FALSE,
  cores = 1) # 911


mas_res_df <- mas_1$results

fdr_mas <- mas_res_df %>%
    dplyr::filter(qval < 0.05) # qval 이면 없다. 

Maaslin2_result_Xfdr <- mas_res_df %>%
    dplyr::filter(pval < 0.05) %>% .$feature

dim(fdr_mas) # Johnsonella_ignava.X87152.
fdr_mas


# mas_2 <- Maaslin2(
#   input_data = data.frame(otu_table(oral.sp2)),
#   input_metadata = data.frame(sample_data(oral.sp2)),
#   output = "./output/5.DA_decontam/etc/Maaslin2_2",
#   min_abundance = 0.0,
#   min_prevalence = 0.0,
#   normalization = "CSS",
#   transform = "LOG",
#   analysis_method = "LM",
#   max_significance = 0.05,
#   fixed_effects = "Group",
#   correction = "BH",
#   standardize = FALSE,
#   cores = 1) # 911
# 
# mas_2_res_df <- mas_2$results
# 
# fdr_mas_2 <- mas_2_res_df %>%
#     dplyr::filter(qval < 0.05)
# 
# dim(fdr_mas_2)
```


## 5) ancombc2 : pg OK But결과를 신뢰할 수가 없다 ..............................
```{r}

ancom_da <- ancombc(phyloseq = oral.sp2, 
                    formula = "Group",
                    p_adj_method = "fdr",
                    # lib_cut = 1000,
                    # zero_cut = 0.90,
                    group = "Group", struc_zero = TRUE, neg_lb = TRUE, tol = 1e-5,
                    max_iter = 100, conserve = TRUE, alpha = 0.05, global = FALSE)


ancom_da$res$diff_abn

ancom_res_df <- data.frame(
  Species = ancom_da$res$diff_abn$taxon,
  # beta = unlist(ancom_da$res$beta),
  se = unlist(ancom_da$res$se),
  W = unlist(ancom_da$res$W),
  p_val = unlist(ancom_da$res$p_val),
  q_val = unlist(ancom_da$res$q_val),
  diff_abn = unlist(ancom_da$res$diff_abn))

fdr_ancom <- ancom_res_df %>%
  dplyr::filter(q_val < 0.05)

ancom_res_df %>%
  dplyr::filter(p_val < 0.05) %>% 
  arrange(Species)


dim(fdr_ancom) # 180
fdr_ancom[fdr_ancom$Species =="Porphyromonas_gingivalis(AB035459)", ]
```

```{r}

ancom_da.kg <- ancombc(phyloseq = oral.sp2.kg, 
                    formula = "Group",
                    p_adj_method = "fdr",
                    # lib_cut = 1000,
                    # zero_cut = 0.90,
                    group = "Group", struc_zero = TRUE, neg_lb = TRUE, tol = 1e-5,
                    max_iter = 100, conserve = TRUE, alpha = 0.05, global = FALSE)


ancom.kg_res_df <- data.frame(
  Species = ancom_da.kg$res$diff_abn$taxon,
  se = unlist(ancom_da.kg$res$se),
  W = unlist(ancom_da.kg$res$W),
  p_val = unlist(ancom_da.kg$res$p_val),
  q_val = unlist(ancom_da.kg$res$q_val),
  diff_abn = unlist(ancom_da.kg$res$diff_abn))

fdr_ancom.kg <- ancom.kg_res_df %>%
  dplyr::filter(q_val < 0.05)

dim(fdr_ancom.kg) # 193
fdr_ancom.kg[fdr_ancom.kg$Species =="Porphyromonas_gingivalis(AB035459)", ]

```
```{r}

ancom_da.bu <- ancombc(phyloseq = oral.sp2.bu, 
                    formula = "Group",
                    p_adj_method = "fdr",
                    # lib_cut = 1000,
                    # zero_cut = 0.90,
                    group = "Group", struc_zero = TRUE, neg_lb = TRUE, tol = 1e-5,
                    max_iter = 100, conserve = TRUE, alpha = 0.05, global = FALSE)


ancom.bu_res_df <- data.frame(
  Species = ancom_da.bu$res$diff_abn$taxon,
  se = unlist(ancom_da.bu$res$se),
  W = unlist(ancom_da.bu$res$W),
  p_val = unlist(ancom_da.bu$res$p_val),
  q_val = unlist(ancom_da.bu$res$q_val),
  diff_abn = unlist(ancom_da.bu$res$diff_abn))

fdr_ancom.bu <- ancom.bu_res_df %>%
  dplyr::filter(q_val < 0.05)

dim(fdr_ancom.bu) # 193
fdr_ancom.bu[fdr_ancom.bu$Species =="Porphyromonas_gingivalis(AB035459)", ]

```



## 6) Aldex2 : X
```{r}
# https://microbiome.github.io/OMA/differential-abundance.html

library(ALDEx2)
library(microbiome)
oral.sp2

?aldex

set.seed(12345)
system.time({
        aldex2_oral = aldex(otu_table(oral.sp2), sample_data(oral.sp2)$Group, 
                           mc.samples = 500, 
                           test = "t", effect = TRUE, 
                           denom = "iqlr", verbose = TRUE)
})

rownames_to_column(aldex2_oral, "Species") %>%
  arrange(wi.eBH)
  filter(wi.eBH <= 0.05)  %>% # here we chose the wilcoxon output rather than tt
  dplyr::select(we.eBH, wi.eBH, effect, overlap)
# XXXXX

  
b <- rownames_to_column(aldex2_oral, "Species") %>%
  arrange(wi.ep) %>%  filter(wi.ep <= 0.05) 

aldex2_result_Xfdr <- b$Species
```

```{r}

system.time({
        aldex2_kg = aldex(otu_table(oral.sp2.kg), sample_data(oral.sp2.kg)$Group, 
                           mc.samples = 500, 
                           test = "t", effect = TRUE, 
                           denom = "iqlr", verbose = TRUE)
})

rownames_to_column(aldex2_kg, "Species") %>%
  arrange(wi.eBH)

system.time({
        aldex2_bu = aldex(otu_table(oral.sp2.bu), sample_data(oral.sp2.bu)$Group, 
                           mc.samples = 500, 
                           test = "t", effect = TRUE, 
                           denom = "iqlr", verbose = TRUE)
})

rownames_to_column(aldex2_bu, "Species") %>%
  arrange(wi.eBH)

```


## 7) lefser
```{r}

# microbiomeMarker
?run_lefse
oral.sp2.rel <-  transform_sample_counts(oral.sp2, function(x) x / sum(x) )
oral.group.sp <- run_lefse(
  oral.sp2.rel,
  wilcoxon_cutoff = 0.05,
  group = "Group",
  kw_cutoff = 0.05,
  taxa_rank = "Species",
  multigrp_strat = TRUE,
  lda_cutoff = 2,
  bootstrap_n = 999
    
)
microbiomeMarker::marker_table(oral.group.sp) # 53 

#####################################lefser

BiocManager::install("lefser")

library(phyloseq)
library(lefser)
library(SummarizedExperiment)

# phyloseq 객체 생성 (예시 데이터 사용)
# 여기서는 otu_table, tax_table, sample_data를 적절한 값으로 대체해야 합니다.

# Lefse 분석을 위한 데이터 준비
oral.sp2.rel <-  transform_sample_counts(oral.sp2, function(x) x / sum(x) )
otu_table <- otu_table(oral.sp2.rel)
sample_data <- sample_data(oral.sp2.rel)

oral.sp2.Se  <- SummarizedExperiment(assays = list(counts = otu_table),
                                     colData = sample_data)

# Lefse 분석 수행
?lefser
res <- lefser(expr = oral.sp2.Se,  
              kruskal.threshold = 0.05,
              wilcox.threshold = 0.05,
              lda.threshold = 1,
              groupCol = "Group",
              # assay = 1L,
              # trim.names = "Species"
              )

# 결과 출력
print(res)






```


## 8) wilcoxon RANK SUM TEST
```{r}

oral.sp2.rel

otu <- otu_table(oral.sp2.rel) %>% data.frame
# tax <- tax_table(oral.sp2.rel) %>% data.frame %>% .$Species
metadata <- sample_data(oral.sp2.rel) %>% data.frame()



o.tab <- merge(metadata, as.data.frame(t(otu)), by ="row.names")
rownames(o.tab) <- o.tab$Row.names
o.tab$Row.names <- NULL
o.tab


p_value <- apply(otu, 1,
                 function(x){wilcox.test(as.numeric(x[o.tab[,"Group"] == "AA"]),
                                         as.numeric(x[o.tab[,"Group"] == "HV"]),
                                         correct = F)$p.value})
p_value[p_value <= 0.05] %>% names %>% sort
p_ad <- p.adjust(p_value, method = "fdr" ) %>% sort
p_ad[p_ad < 0.05]
# Johnsonella_ignava(X87152) 
#                 0.02373799 

```

```{r}



otu <- otu_table(oral.sp2.kg.rel) %>% data.frame
tax <- tax_table(oral.sp2.kg.rel) %>% data.frame %>% select(Species)
metadata <- sample_data(oral.sp2.kg.rel) %>% data.frame()


table <- merge(otu, tax, by = "row.names") %>% t() %>% data.frame
colnames(table) <- table["Species",] 
table <- table[-1,]
o.tab <- merge(metadata, table, by ="row.names")
rownames(o.tab) <- o.tab$Row.names
o.tab$Row.names <- NULL
o.tab


p_value <- apply(table, 2,
                 function(x){wilcox.test(as.numeric(x[o.tab[,"Group"] == "AA"]),
                                         as.numeric(x[o.tab[,"Group"] == "HV"]),
                                         correct = F)$p.value})

p_ad <- p.adjust(p_value, method = "fdr" ) %>% sort
p_ad[p_ad < 0.05]



```





## 7) wilcoxon RANK SUM TEST : CLR abundance
```{r}
oral.sp2.clr <- microbiome::transform(oral.sp2, 'clr')

otu <- otu_table(oral.sp2.clr) %>% data.frame
# tax <- tax_table(oral.sp2.clr) %>% data.frame %>% select(Species)
metadata <- sample_data(oral.sp2.clr) %>% data.frame()

o.tab <- merge(metadata, as.data.frame(t(otu)), by ="row.names")
rownames(o.tab) <- o.tab$Row.names
o.tab$Row.names <- NULL
o.tab


p_value <- apply(otu, 1,
                 function(x){wilcox.test(as.numeric(x[o.tab[,"Group"] == "AA"]),
                                         as.numeric(x[o.tab[,"Group"] == "HV"]),
                                         correct = F)$p.value})

p_ad <- p.adjust(p_value, method = "fdr" ) %>% sort %>% data.frame() %>% rownames_to_column("Species")
p_ad[p_ad$. < 0.05,"Species"] %>% sort# 312  
# Wilcox_CLR <- p_ad[p_ad$. < 0.05,"Species"] %>% sort# 312  
p_ad[p_ad$Species == "Porphyromonas_gingivalis(AB035459)", ] # OK


 
Wilcox_CLR <- 
```


```{r}
oral.sp2.bu.clr <- microbiome::transform(oral.sp2.bu, 'clr')

otu <- otu_table(oral.sp2.bu.clr) %>% data.frame
tax <- tax_table(oral.sp2.bu.clr) %>% data.frame %>% select(Species)
metadata <- sample_data(oral.sp2.bu.clr) %>% data.frame()


table <- merge(otu, tax, by = "row.names") %>% t() %>% data.frame
colnames(table) <- table["Species",] 
table <- table[-1,]
o.tab <- merge(metadata, table, by ="row.names")
rownames(o.tab) <- o.tab$Row.names
o.tab$Row.names <- NULL
o.tab


p_value <- apply(table, 2,
                 function(x){wilcox.test(as.numeric(x[o.tab[,"Group"] == "AA"]),
                                         as.numeric(x[o.tab[,"Group"] == "HV"]),
                                         correct = F)$p.value})

p_ad <- p.adjust(p_value, method = "fdr" ) %>% sort %>% data.frame() %>% rownames_to_column("Species")
p_ad[p_ad$. < 0.05,] 

p_ad[p_ad$Species == "Porphyromonas_gingivalis(AB035459)", ] # OK
```



## 8) t.test.rel


```{r}

oral.sp2.rel

otu <- otu_table(oral.sp2.rel) %>% data.frame
metadata <- sample_data(oral.sp2.rel) %>% data.frame()

o.tab <- merge(metadata, as.data.frame(t(otu)), by ="row.names")
rownames(o.tab) <- o.tab$Row.names
o.tab$Row.names <- NULL
o.tab


p_value <- apply(otu, 1,
                 function(x){t.test(as.numeric(x[o.tab[,"Group"] == "AA"]),
                                         as.numeric(x[o.tab[,"Group"] == "HV"]),
                                         correct = F)$p.value})
p_ad <- p.adjust(p_value, method = "fdr" ) %>% sort
p_ad[p_ad < 0.05] 
t.test_result_Xfdr <- p_value[p_value<=0.05] %>% names %>% sort
```

```{r}



otu <- otu_table(oral.sp2.bu.rel) %>% data.frame
tax <- tax_table(oral.sp2.bu.rel) %>% data.frame %>% select(Species)
metadata <- sample_data(oral.sp2.bu.rel) %>% data.frame()


table <- merge(otu, tax, by = "row.names") %>% t() %>% data.frame
colnames(table) <- table["Species",] 
table <- table[-1,]
o.tab <- merge(metadata, table, by ="row.names")
rownames(o.tab) <- o.tab$Row.names
o.tab$Row.names <- NULL
o.tab


p_value <- apply(table, 2,
                 function(x){t.test(as.numeric(x[o.tab[,"Group"] == "AA"]),
                                         as.numeric(x[o.tab[,"Group"] == "HV"]),
                                         correct = F)$p.value})

p_ad <- p.adjust(p_value, method = "fdr" ) %>% sort
p_ad[p_ad < 0.05]



```


# 5. Vendiagram
- 시도해본 분석 방법
  - aldex2
  - ANCOMBC(ANCOM-II은 R에서 X)
  - DESeq2
  - edgeR
  - Lefser
  - MaAsLin2
  - metagenomeSeq
  - limma voom
  - wilconxon test
  - t test

-  marker 유전자가 발견된 분석 방법은
  - ANCOMBC
  - DESeq2
  - edgeR
  - Lefser
  - Manslin2
  - limma voom
  - Wilcoxon (CLR)


- 우리가 주목하고 싶은 bacteria
  - Porphyromonas gingivalis(AB035459)
  - Tannerella forsythia(AB035460)
  - Treponema denticola(AF139203)

```{r}

# 결과 저장
DESeq2_result <- fdr_deseq$Species
MaAsLin2_result <- fdr_mas$feature
ANCOM_bc_result <- fdr_ancom$Species
edgeR_result <- marker_table(Oral_edger)$feature
lefse_result <- marker_table(oral.group.sp)$feature 
Wilcox_CLR 



# Species이름 맨 뒤에 공백 지우기
DESeq2_result <- trimws(DESeq2_result)
# MaAsLin2_result <-trimws(MaAsLin2_result)
ANCOM_bc_result <-trimws(ANCOM_bc_result)
edgeR_result <- trimws(edgeR_result)
lefse_result <- trimws(lefse_result)
Wilcox_CLR <- trimws(Wilcox_CLR)
# MaAsLin2_result .을 ()로 대체하기 

original_string <- MaAsLin2_result

modified_string1 <- sub("\\.", "(", original_string)
modified_string2 <- sub("\\.(?=[^.]*$)", ")", modified_string1, perl = TRUE)
MaAsLin2_result <- modified_string2

# 
DESeq2_result
ANCOM_bc_result
edgeR_result
lefse_result
Wilcox_CLR

intersect(edgeR_result, lefse_result)    # 20 pg O

intersect(DESeq2_result, ANCOM_bc_result)    # 19 pg O


intersect(ANCOM_bc_result, Wilcox_CLR)    # 11 pg O



# Red complex
redcom <- c("Porphyromonas_gingivalis(AB035459)", 
  "Tannerella_forsythia(AB035460)",
  "Treponema_denticola(AF139203)")
 
redcom %in% DESeq2_result
redcom %in% MaAsLin2_result
redcom %in% lefse_result
redcom %in% ANCOM_bc_result # Treponema_denticola(AF139203) 도 있음ㅇㅇ 
redcom %in% edgeR_result
redcom %in%Wilcox_CLR
```

```{r}
library(VennDiagram)
library(RColorBrewer)


myCol <- c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#E7298A',0.3), alpha('#D95F02',0.3))
# Chart

venn.diagram(
        x = list(DESeq2_result,
                 ANCOM_bc_result,
                 edgeR_result,
                 lefse_result,
                 Wilcox_CLR),
        category.names = c(# "Limma-Voom" , 
                           "DESeq2",
                           "ANCOM-BC",
                           "edgeR",
                           "lefse",
                           "Wilcoxon(clr)"),
        filename = './output/5.DA_decontam/etc/venn_diagramm.png',
        output=TRUE,
          col=c("#440154ff", '#21908dff', '#fde725ff', "#E7298A", "#D95F02"),
          height = 10,
          width = 10, 
          resolution = 300,
          units = "in",
          imagetype = "png",
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#E7298A',0.3), alpha('#D95F02',0.3)),
          cex = 1,
          fontfamily = "sans",
          cat.cex = 1,
          cat.pos = c(0, 300, 200, 160, 2),
          cat.default.pos = "outer",
          cat.fontfamily = "sans",
          cat.col = c("#440154ff", '#21908dff', '#fde725ff', "#E7298A", "#D95F02"),
)


```

## Before FDR
```{r}

limma_result_Xfdr
Corncob_result_Xfdr


aldex2_result_Xfdr
Maaslin2_result_Xfdr
metaseq_result_Xfdr
t.test_result_Xfdr


venn.diagram(
        x = list(aldex2_result_Xfdr,
                 Maaslin2_result_Xfdr,
                 metaseq_result_Xfdr,
                 t.test_result_Xfdr),
        category.names = c("Aldex2",
                           "Maaslin2",
                           "MetagenomeSeq",
                           "T-test"),
        filename = './output/5.DA_decontam/etc/venn_diagramm_xFDR.png',
        output=TRUE,
          col=c("#440154ff", '#21908dff', '#fde725ff', "#E7298A" ),
          height = 10,
          width = 10, 
          resolution = 300,
          units = "in",
          imagetype = "png",
          fill = c(alpha("#440154ff",0.3), alpha('#21908dff',0.3), alpha('#fde725ff',0.3), alpha('#E7298A',0.3)),
          cex = 1,
          fontfamily = "sans",
          cat.cex = 1,
          cat.default.pos = "outer",
          cat.fontfamily = "sans",
          cat.col = c("#440154ff", '#21908dff', '#fde725ff', "#E7298A"),
)




```


# 